File: DemoProject/kakao-ios-sdk-1/samples/objc/KakaoSample/KakaoSample/TalkProfileViewController.m
/**
 * Copyright 2015-2018 Kakao Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "TalkProfileViewController.h"
#import "TalkProfileTableViewCell.h"
#import "ProfileImageViewController.h"
#import "UIAlertController+Addition.h"
#import <KakaoOpenSDK/KakaoOpenSDK.h>

@interface TalkProfileViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property (strong, nonatomic) KOTalkProfile *profile;
@property (strong, nonatomic) UITapGestureRecognizer *singleTapGesture;

@end

@implementation TalkProfileViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _singleTapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(profileImageTapped:)];
    _singleTapGesture.numberOfTapsRequired = 1;
    _singleTapGesture.numberOfTouchesRequired = 1;
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.tableFooterView = [[UIView alloc] init];
    
    UINib *nib = [UINib nibWithNibName:@"TalkProfileTableViewCell" bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:@"TalkProfileTableViewCell"];
    
    [KOSessionTask talkProfileTaskWithCompletionHandler:^(id result, NSError *error) {
        if (error) {
            [UIAlertController showMessage:error.description];
        } else {
            self.profile = result;
            [self.tableView reloadData];
        }
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TalkProfileTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TalkProfileTableViewCell"];
    [cell setTalkProfile:_profile];
    
    BOOL hasGesture = NO;
    for (UIGestureRecognizer *gesture in [cell.thumbnail gestureRecognizers]) {
        if (gesture == _singleTapGesture) {
            hasGesture = YES;
            break;
        }
    }
    
    if (!hasGesture) {
        [cell.thumbnail addGestureRecognizer:_singleTapGesture];
        cell.thumbnail.userInteractionEnabled = YES;
    }
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 236;
}

- (void)profileImageTapped:(UITapGestureRecognizer *)recognizer {
    if (!_profile || _profile.profileImageURL.length == 0) {
        return;
    }
    
    [self performSegueWithIdentifier:@"ProfileImage" sender:_profile.profileImageURL];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"ProfileImage"]) {
        ProfileImageViewController *viewController = segue.destinationViewController;
        viewController.profileImageUrlString = sender;
    }
}

@end

File: DemoProject/kakao-ios-sdk-1/samples/swift/KakaoSample/KakaoSample/TalkProfileViewController.swift
/**
 * Copyright 2015-2018 Kakao Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

class TalkProfileViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!
    
    fileprivate var profile:KOTalkProfile?
    fileprivate var singleTapGesture: UITapGestureRecognizer!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        singleTapGesture = UITapGestureRecognizer(target: self, action: #selector(TalkProfileViewController.profileImageTapped(_:)))
        singleTapGesture.numberOfTapsRequired = 1
        singleTapGesture.numberOfTouchesRequired = 1
        
        tableView.delegate = self;
        tableView.dataSource = self
        tableView.tableFooterView = UIView(frame: CGRect.zero)
        
        let nib = UINib(nibName: "TalkProfileTableViewCell", bundle: nil)
        tableView.register(nib, forCellReuseIdentifier: "TalkProfileTableViewCell")
        
        KOSessionTask.talkProfileTask { (talkProfile, error) -> Void in
            if error != nil {
                UIAlertController.showMessage("에러! \(error!)")
            } else {
                self.profile = (talkProfile as! KOTalkProfile)
                self.tableView.reloadData()
            }
        }
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "TalkProfileTableViewCell") as! TalkProfileTableViewCell
        cell.setTalkProfile(self.profile)
        
        var addGesture = true
        if let gestures: [AnyObject] = cell.thumbnail.gestureRecognizers {
            addGesture = gestures.filter {$0 as? UIGestureRecognizer == self.singleTapGesture}.isEmpty
        }
        
        if addGesture {
            cell.thumbnail.addGestureRecognizer(singleTapGesture)
            cell.thumbnail.isUserInteractionEnabled = true
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 236
    }

    @objc func profileImageTapped(_ recognizer: UITapGestureRecognizer) {
        if profile == nil || profile?.profileImageURL == nil {
            return
        }
        
        if let imageUrl = profile?.profileImageURL! , imageUrl.isEmpty {
            return
        }
        
        self.performSegue(withIdentifier: "ProfileImage", sender: profile?.profileImageURL)
    }
    
    // MARK: - Navigation
    
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "ProfileImage" {
            let viewController: ProfileImageViewController = (segue.destination as? ProfileImageViewController)!
            viewController.profileImageUrlString = sender as? String
            
        }
    }

}
